sound/soc/sof/ipc4-pcm.c

```diff
@@ -495,17 +496,36 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 				use_chain_dma = true;
 		}
 	}
+
+	fe_rate = params_rate(params);
+
 	/*
-	 * Always set BE format to 32-bits for both playback and capture, except if chained
-	 * DMA is in use. In chained DMA case the back-end and front-end format must match.
+	 * Limit sampling rate if copier has a format that matches
+	 * the rate.
 	 */
 	if (!use_chain_dma) {
-		snd_mask_none(fmt);
-		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+		struct sof_ipc4_available_audio_format *available_fmt = &ipc4_copier->available_fmt;
+		int i;
+
+		for (i = 0; i < available_fmt->num_output_formats; i++) {
+			struct sof_ipc4_audio_format *audio_fmt =
+				&available_fmt->output_pin_fmts[i].audio_fmt;
+
+			if (fe_rate == audio_fmt->sampling_frequency)
+				fe_be_rate_match = true;
```

Comment:
break out from the loop here?

Reply:
you can still break out from here...
the `single_be_rate` does not matter if `fe_be_rate_match` is true.